### PR TITLE
Align permission API's with Java/Swift

### PR DESCRIPTION
**Pull request by cmelchior**
This PR aligns the JS API with Java/Swift by introducing a number of helper methods.

It adds
```
Realm.permissions()
Realm.permissions(type)
Realm.Permissions.Realm.findOrCreate(roleName)
Realm.Permissions.Class.findOrCreate(roleName)
```

TODO:
* [x] Fix changelog
* [x] Add `permissions()` method + test
* [x] Add `findOrCreate(roleName)` + tests

In order for Realm to pick up extra methods on model classes we have to register the Javascript constructor function with Realm. This, unfortunately, means I have to revert the change by Nabil and move it back to JS land, but it is done in a way compatible with the C++ implementation so that `new Realm(config)` will also include the schema. This is done by adding the internal `_constructor` method as wrapping the original constructor itself seemed to cause other places in our code base (and  possible client code as well).

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for review

File: docs/permission.js [hunk 1]
```diff
@@ -370,14 +370,37 @@ class Class {
...
     get permissions() {}
+
+    /**
+     * Finds the Class-level permissions associated with the named Role. If either the role or the permission
+     * object doesn't exists, it will be created.
+     *
+     * If the Permission object is created because one didn't exist already, it will be
+     * created with all privileges disabled.
+     *
+     * If the Role object is created because one didn't exists, it will be created
```
Comment (kneth) [inline docs/permission.js]: "exists" -> "exist"

File: docs/permission.js [hunk 2]
```diff
@@ -370,14 +370,37 @@ class Class {
...
+
     /**
      * The permissions for this class.
      * @type {Array<Realm.Permissions.Permission>}
      */
     get permissions() {}
+
+    /**
+     * Finds the Class-level permissions associated with the named Role. If either the role or the permission
+     * object doesn't exists, it will be created.
```
Comment (kneth) [inline docs/permission.js]: "exists" -> "exist"

File: docs/permission.js [hunk 3]
```diff
@@ -389,9 +412,25 @@ class Class {
...
+
     /**
      * The permissions for the Realm.
      * @type {Array<Realm.Permissions.Permission>}
      */
     get permissions() {}
+
+    /**
+     * Finds the Realm-level permissions associated with the named Role. If either the role or the permission
+     * object doesn't exists, it will be created.
```
Comment (kneth) [inline docs/permission.js]: "exists" -> "exist"

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,8 +1,13 @@
 x.x.x Release notes (yyyy-MM-dd)
 =============================================================
 ## Enhancements
-* None
-
+* Added support for finding Realm-level permissions in Query-based Realms using `realm.getPermissions()` ([#2036]
```
Comment (kneth) [inline CHANGELOG.md]: Please move `.` to end of sentence: "Added ... using `realm.getPermissions()`. ([..."

File: docs/realm.js [hunk 5]
```diff
@@ -174,6 +174,17 @@ class Realm {
      */
     privileges(arg) {}
 
+    /**
+     * Returns the fine-grained permissions object associated with either the Realm itself or a Realm model class.
+     *
+     * @param {Realm~ObjectType} [arg] - If no argument is provided, the Realm-level permissions are are returned.
```
Comment (kneth) [inline docs/realm.js]: "are are" -> "are"

File: lib/extensions.js [hunk 6]
```diff
@@ -285,31 +322,99 @@ module.exports = function(realmConstructor) {
...
+                    if (classPermissions.length === 0) {
+                        throw Error(`Could not find Class-level permissions for '${schemaName}'`);
+                    }
+                    return classPermissions[0];
+                }
+            },
+        }));
     }
 
+    // Realm instance methods that are available always available
```
Comment (kneth) [inline lib/extensions.js]: Remove one "available"

**Comment by bmunkholm** [comment]
Changelog is fine.

**Pull request merged**